modules/kibana/src/internalClusterTest/java/org/elasticsearch/kibana/KibanaThreadPoolIT.java

```diff
@@ -61,4 +97,108 @@ public void testKibanaThreadPool() {
     }
+
+    public void testBlockedThreadPoolsRejectUserRequests() throws Exception {
+        assertAcked(client().admin().indices().prepareCreate(USER_INDEX));
+
+        runWithBlockedThreadPools(this::assertThreadPoolsBlocked);
+
+        assertAcked(client().admin().indices().prepareDelete(USER_INDEX));
+    }
+
+    private void assertThreadPoolsBlocked() {
+
+        var e1 = expectThrows(
+            EsRejectedExecutionException.class,
+            () -> client().prepareIndex(USER_INDEX).setSource(Map.of("foo", "bar")).get()
+        );
+        assertThat(e1.getMessage(), startsWith("rejected execution of TimedRunnable"));
+        var e2 = expectThrows(EsRejectedExecutionException.class, () -> client().prepareGet(USER_INDEX, "id").get());
+        assertThat(e2.getMessage(), startsWith("rejected execution of ActionRunnable"));
+        var e3 = expectThrows(
+            SearchPhaseExecutionException.class,
+            () -> client().prepareSearch(USER_INDEX)
+                .setQuery(QueryBuilders.matchAllQuery())
+                // Request times out if max concurrent shard requests is set to 1
+                .setMaxConcurrentShardRequests(usually() ? SearchRequest.DEFAULT_MAX_CONCURRENT_SHARD_REQUESTS : randomIntBetween(2, 10))
+                .get()
+        );
+        assertThat(e3.getMessage(), containsString("all shards failed"));
+    }
+
+    protected void runWithBlockedThreadPools(Runnable runnable) throws Exception {
+        Phaser phaser = new Phaser();
+
+        // register this test's thread
+        phaser.register();
+
+        blockThreadPool(phaser);
+        phaser.arriveAndAwaitAdvance();
+
+        fillQueues();
+
+        logger.debug("number of nodes " + internalCluster().getNodeNames().length);
+        logger.debug("number of parties arrived " + phaser.getArrivedParties());
+        try {
+            runnable.run();
+        } finally {
+            phaser.arriveAndAwaitAdvance();
+        }
+    }
+
+    private void blockThreadPool(Phaser phaser) {
+        for (String nodeName : internalCluster().getNodeNames()) {
+            ThreadPool threadPool = internalCluster().getInstance(ThreadPool.class, nodeName);
+            for (String threadPoolName : THREAD_POOLS_TO_BLOCK) {
+                blockThreadPool(threadPoolName, threadPool, phaser);
+            }
+        }
+    }
+
+    private void fillQueues() {
+        for (String nodeName : internalCluster().getNodeNames()) {
+            ThreadPool threadPool = internalCluster().getInstance(ThreadPool.class, nodeName);
+            for (String threadPoolName : THREAD_POOLS_TO_BLOCK) {
+                fillThreadPoolQueues(threadPoolName, threadPool);
+            }
+        }
+    }
+
+    private static void blockThreadPool(String threadPoolName, ThreadPool threadPool, Phaser phaser) {
+        ThreadPool.Info info = threadPool.info(threadPoolName);
+
+        Runnable waitAction = () -> {
+            phaser.arriveAndAwaitAdvance();
```

Comment:
Nit: maybe add a comment to indicate which step each of the arriveAndAwait guards